src/server/project.ts

```diff
@@ -763,16 +749,21 @@ namespace ts.server {
         }
 
         /* @internal */
-        private extractUnresolvedImportsFromSourceFile(file: SourceFile, result: Push<string>, ambientModules: string[]) {
+        private extractUnresolvedImportsFromSourceFile(file: SourceFile, result: string[] | undefined, ambientModules: string[]): string[] | undefined {
+            // No unresolve imports in this file
+            if (this.filesWithNoUnresolvedImports.has(file.path)) {
+                return result;
+            }
+
             const cached = this.cachedUnresolvedImportsPerFile.get(file.path);
             if (cached) {
                 // found cached result - use it and return
                 for (const f of cached) {
-                    result.push(f);
+                    (result || (result = [])).push(f);

```

Comment:
kinda weird that we take result as an input and return it as an output.. i would just do one of them..